packages/alfa-rules/src/sia-r69/rule.ts

```diff
@@ -0,0 +1,388 @@
+import { expectation } from "../common/expectation";
+
+import { hasCategory } from "../common/predicate/has-category";
+import { hasRole } from "../common/predicate/has-role";
+import { isDisabled } from "../common/predicate/is-disabled";
+import { isIgnored } from "../common/predicate/is-ignored";
+import { isVisible } from "../common/predicate/is-visible";
+
+import { Question } from "../common/question";
+
+const { reduce, some, flatMap, map, concat } = Iterable;
+const { and, not, equals, test } = Predicate;
+const { min, max } = Math;
+
+export default Rule.Atomic.of<Page, Text, Question>({
+  uri: "https://siteimprove.github.io/sanshikan/rules/sia-r69.html",
+  evaluate({ device, document }) {
+    return {
+      applicability() {
+        return visit(document);
+
+        function* visit(node: Node): Iterable<Text> {
+          if (Element.isElement(node)) {
+            if (
+              test(Element.hasNamespace(not(equals(Namespace.HTML))), node) ||
+              test(hasRole(hasCategory(equals(Role.Category.Widget))), node) ||
+              test(
+                and(hasRole(Role.hasName(equals("group"))), isDisabled),
+                node
+              )
+            ) {
+              return;
+            }
+          }
+
+          if (Text.isText(node)) {
+            if (test(and(isVisible(device), not(isIgnored(device))), node)) {
+              yield node;
+            }
+          }
+
+          const children = node.children({ flattened: true, nested: true });
+
+          for (const child of children) {
+            yield* visit(child);
+          }
+        }
+      },
+
+      expectations(target) {
+        const foregrounds = Question.of(
+          "foreground-colors",
+          "color[]",
+          target,
+          "What are the foreground colors of the text node?"
+        );
+
+        const backgrounds = Question.of(
+          "background-colors",
+          "color[]",
+          target,
+          "What are the background colors of the text node?"
+        );
+
+        const result = foregrounds.map((foregrounds) =>
+          backgrounds.map((backgrounds) => {
+            const contrasts = flatMap(foregrounds, (foreground) =>
+              map(backgrounds, (background) => contrast(foreground, background))
+            );
+
+            const highest = reduce(
+              contrasts,
+              (lowest, contrast) => max(lowest, contrast),
+              0
+            );
+
+            const threshold = isLargeText(device)(target) ? 3 : 4.5;
+
+            return expectation(
+              highest >= threshold,
+              () => Outcomes.HasSufficientContrast,
+              () => Outcomes.HasInsufficientContrast
+            );
+          })
+        );
+
+        const parent = target.parent({ flattened: true }).get() as Element;
+
+        return {
+          1: getForeground(parent, device)
+            .map((foreground) => result.answer(foreground))
+            .flatMap((result) =>
```

Comment:
Naming the local variable `result`, same as the global(-ish) variable that is used both on the previous line and after this call (l109) is rather confusing…

Reply:
Let me know if https://github.com/Siteimprove/alfa/pull/228#discussion_r420061683 provides some clarity.